esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5885,4 +5892,260 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+             * Not least because IDF allows updates via IPT -> Dali.
+             *
+             * So for now don't send anything, and rely on engine fetching the legacy way, i.e. direct from Dali, via lazy fetching etc.
+             */
+            break;
+        }
+        case CFileAccessRole_External:
+        {
+            IArrayOf<IEspDFUPartLocations> dfuPartLocations;
+            IArrayOf<IEspDFUPartCopies> dfuPartCopies;
+            resp.setNumParts(getFilePartsInfo(context, df, req.getCluster(), dfuPartLocations, dfuPartCopies));
+            resp.setFilePartLocations(dfuPartLocations);
+            resp.setFileParts(dfuPartCopies);
+            if (req.getReturnJsonTypeInfo() || req.getReturnJsonTypeInfo())
+            {
+                MemoryBuffer binLayout;
+                StringBuffer jsonLayout;
+                if (!getRecordFormatFromRtlType(binLayout, jsonLayout, df->queryAttributes(), req.getReturnJsonTypeInfo(), req.getReturnJsonTypeInfo()))
+                    getRecordFormatFromECL(binLayout, jsonLayout, df->queryAttributes(), req.getReturnJsonTypeInfo(), req.getReturnJsonTypeInfo());
+                if (req.getReturnJsonTypeInfo() && jsonLayout.length())
+                    resp.setRecordTypeInfoJson(jsonLayout.str());
+                if (req.getReturnBinTypeInfo() && binLayout.length())
+                    resp.setRecordTypeInfoBin(binLayout);
+            }
+            break;
+        }
+        default:
+            throwUnexpected();
+    }
+
+    // setup "expiryTime"
+    unsigned expirySecs = req.getExpirySeconds();
+    if (expirySecs > fileAccessExpirySeconds)
+        expirySecs = fileAccessExpirySeconds;
+    time_t now;
+    time(&now);
+    CDateTime expiryDt;
+    expiryDt.set(now + expirySecs);
+    StringBuffer expiryTimeStr;
+    expiryDt.getString(expiryTimeStr);
+
+    StringBuffer keyPairName;
+    getFileDafilesrvKeyName(keyPairName, *df);
+
+    Owned<IPropertyTree> metaInfo = createPTree();
+    getFileMeta(*metaInfo, *df, udesc, role, expiryTimeStr, keyPairName, req);
+    StringBuffer metaInfoJsonBlob;
+    toJSON(metaInfo, metaInfoJsonBlob);
+    resp.setMetaInfoBlob(metaInfoJsonBlob);
+    resp.setKeyName(keyPairName);
+    resp.setExpiryTime(expiryTimeStr);
+}
+
+
+SecAccessFlags translateToSecAccessFlags(CSecAccessType from)
+{
+    switch (from)
+    {
+        case CSecAccessType_Access:
+            return SecAccess_Access;
+        case CSecAccessType_Read:
+            return SecAccess_Read;
+        case CSecAccessType_Write:
+            return SecAccess_Write;
+        case CSecAccessType_Full:
+            return SecAccess_Full;
+        case CSecAccessType_None:
+        default:
+            return SecAccess_None;
+    }
+}
+
+bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &req, IEspDFUFileAccessResponse &resp)
+{
+    try
+    {
+        SecAccessFlags accessType = translateToSecAccessFlags(req.getAccessType());
```

Comment:
Not sure why you need the translation here. Why not use CSecAccessType?

Reply:
well yes you could, but it would have been better if CSecAccessType was not necessary at all.
I'm guessing it had to be introduced because you couldn't use the pre-existing enumberated type which this mirror (i.e. SecAccessFlags) in ecm?
If you could have used SecAccessFlags that would have been better than introducing a clone of it.

So it's best to limit the use of this type that was only necessary for ECM to only where absolutely required and make other methods use the generic security access type we use elsewhere throughout the platform.